### ISSUE TITLE
issues with docker running server

### ISSUE THREAD

**Issue by bcolloran** [iodide-project/iodide#716]
hi @wlach i tried to get the server up to test it, but hit some kind of snag. here is my console output:

```
docker-compose up
Creating network "iodide_default" with the default driver
Pulling db (postgres:9.5-alpine)...
9.5-alpine: Pulling from library/postgres
8e3ba11ec2a2: Already exists
fbb9adb8cff8: Pull complete
aa8817b9e00d: Pull complete
4bf2aaaa6fa5: Pull complete
f2d9b733e67e: Pull complete
c080b0dfa3cb: Pull complete
156656f194bd: Pull complete
d63dd09852f1: Pull complete
91fd29ed4197: Pull complete
Digest: sha256:f60d9ae8516349ad174ab8e99b40612bdf19c23c10cd8f633cb50bb93627b177
Status: Downloaded newer image for postgres:9.5-alpine
Creating iodide_db_1 ... done
Creating iodide_server_1 ... done
Attaching to iodide_db_1, iodide_server_1
db_1      | WARNING: no logs are available with the 'none' log driver
server_1  | Traceback (most recent call last):
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 216, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 194, in connect
server_1  |     self.connection = self.get_new_connection(conn_params)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/postgresql/base.py", line 168, in get_new_connection
server_1  |     connection = Database.connect(**conn_params)
server_1  |   File "/usr/local/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
server_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
server_1  | psycopg2.OperationalError: could not connect to server: Connection refused
server_1  | 	Is the server running on host "db" (172.18.0.2) and accepting
server_1  | 	TCP/IP connections on port 5432?
server_1  | 
server_1  | 
server_1  | The above exception was the direct cause of the following exception:
server_1  | 
server_1  | Traceback (most recent call last):
server_1  |   File "manage.py", line 10, in <module>
server_1  |     execute_from_command_line(sys.argv)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
server_1  |     utility.execute()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 365, in execute
server_1  |     self.fetch_command(subcommand).run_from_argv(self.argv)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/core/management/base.py", line 288, in run_from_argv
server_1  |     self.execute(*args, **cmd_options)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/core/management/base.py", line 335, in execute
server_1  |     output = self.handle(*args, **options)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/core/management/commands/migrate.py", line 79, in handle
server_1  |     executor = MigrationExecutor(connection, self.migration_progress_callback)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/migrations/executor.py", line 18, in __init__
server_1  |     self.loader = MigrationLoader(self.connection)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/migrations/loader.py", line 49, in __init__
server_1  |     self.build_graph()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/migrations/loader.py", line 207, in build_graph
server_1  |     self.applied_migrations = recorder.applied_migrations()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/migrations/recorder.py", line 61, in applied_migrations
server_1  |     if self.has_table():
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/migrations/recorder.py", line 44, in has_table
server_1  |     return self.Migration._meta.db_table in self.connection.introspection.table_names(self.connection.cursor())
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 255, in cursor
server_1  |     return self._cursor()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 232, in _cursor
server_1  |     self.ensure_connection()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 216, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/utils.py", line 89, in __exit__
server_1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 216, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/base/base.py", line 194, in connect
server_1  |     self.connection = self.get_new_connection(conn_params)
server_1  |   File "/usr/local/lib/python3.7/site-packages/django/db/backends/postgresql/base.py", line 168, in get_new_connection
server_1  |     connection = Database.connect(**conn_params)
server_1  |   File "/usr/local/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
server_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
server_1  | django.db.utils.OperationalError: could not connect to server: Connection refused
server_1  | 	Is the server running on host "db" (172.18.0.2) and accepting
server_1  | 	TCP/IP connections on port 5432?
server_1  | 
iodide_server_1 exited with code 1

```

**Comment by wlach** [comment]
@bcolloran So the recommended set of steps for bringing up the server are documented here:

https://github.com/iodide-project/iodide#server-mode

I suspect the problem is that you didn't run the `cp .env-dist .env` step.

**Comment by wlach** [comment]
I'm going to close this for now, please reopen if you still can't get it working after running through the above set of steps.

**Comment by nrgoncalves** [comment]
Hello @wlach, just to report that I had a somewhat similar issue when setting up a local dev environment as detailed in

https://iodide-project.github.io/docs/local-dev/

The issue seems to be that the database is not ready by the time the migrations start:

```
server_1  | Traceback (most recent call last):
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 220, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 197, in connect
server_1  |     self.connection = self.get_new_connection(conn_params)
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/postgresql/base.py", line 185, in get_new_connection
server_1  |     connection = Database.connect(**conn_params)
server_1  |   File "/venv/lib/python3.8/site-packages/psycopg2/__init__.py", line 127, in connect
server_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
server_1  | psycopg2.OperationalError: could not connect to server: Connection refused
server_1  | 	Is the server running on host "db" (172.22.0.3) and accepting
server_1  | 	TCP/IP connections on port 5432?
server_1  | 
server_1  | 
server_1  | The above exception was the direct cause of the following exception:
server_1  | 
server_1  | Traceback (most recent call last):
server_1  |   File "manage.py", line 10, in <module>
server_1  |     execute_from_command_line(sys.argv)
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
server_1  |     utility.execute()
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
server_1  |     self.fetch_command(subcommand).run_from_argv(self.argv)
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/base.py", line 328, in run_from_argv
server_1  |     self.execute(*args, **cmd_options)
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/base.py", line 369, in execute
server_1  |     output = self.handle(*args, **options)
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/base.py", line 83, in wrapped
server_1  |     res = handle_func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 86, in handle
server_1  |     executor = MigrationExecutor(connection, self.migration_progress_callback)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/migrations/executor.py", line 18, in __init__
server_1  |     self.loader = MigrationLoader(self.connection)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/migrations/loader.py", line 49, in __init__
server_1  |     self.build_graph()
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/migrations/loader.py", line 212, in build_graph
server_1  |     self.applied_migrations = recorder.applied_migrations()
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/migrations/recorder.py", line 76, in applied_migrations
server_1  |     if self.has_table():
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/migrations/recorder.py", line 56, in has_table
server_1  |     return self.Migration._meta.db_table in self.connection.introspection.table_names(self.connection.cursor())
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 260, in cursor
server_1  |     return self._cursor()
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 236, in _cursor
server_1  |     self.ensure_connection()
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 220, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
server_1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 220, in ensure_connection
server_1  |     self.connect()
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 197, in connect
server_1  |     self.connection = self.get_new_connection(conn_params)
server_1  |   File "/venv/lib/python3.8/site-packages/django/utils/asyncio.py", line 26, in inner
server_1  |     return func(*args, **kwargs)
server_1  |   File "/venv/lib/python3.8/site-packages/django/db/backends/postgresql/base.py", line 185, in get_new_connection
server_1  |     connection = Database.connect(**conn_params)
server_1  |   File "/venv/lib/python3.8/site-packages/psycopg2/__init__.py", line 127, in connect
server_1  |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
server_1  | django.db.utils.OperationalError: could not connect to server: Connection refused
server_1  | 	Is the server running on host "db" (172.22.0.3) and accepting
server_1  | 	TCP/IP connections on port 5432?
db_1      |  done
db_1      | server started
db_1      | 
db_1      | /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
db_1      | 
db_1      | waiting for server to shut down....LOG:  received fast shutdown request
db_1      | LOG:  aborting any active transactions
db_1      | LOG:  autovacuum launcher shutting down
db_1      | LOG:  shutting down
db_1      | LOG:  database system is shut down
iodide_server_1 exited with code 1
db_1      |  done
db_1      | server stopped
db_1      | 
db_1      | PostgreSQL init process complete; ready for start up.
db_1      | 
db_1      | LOG:  database system was shut down at 2020-08-02 20:59:37 UTC
db_1      | LOG:  MultiXact member wraparound protections are now enabled
db_1      | LOG:  autovacuum launcher started
db_1      | LOG:  database system is ready to accept connections

```

A quick and dirty change to `bin/run` seems to avoid the issue:

```
case $1 in
  dev)
    sleep 60
    python manage.py migrate --noinput
    exec python manage.py runserver 0.0.0.0:${PORT}
    ;;
```

**Comment by wlach** [comment]
@nrgoncalves yeah I think there's a race in the docker-compose code, we have a workaround for ci:

https://github.com/iodide-project/iodide/blob/5c38c734420edb37ccff5ea3243d6125a498be8c/.circleci/config.yml#L140

... but I haven't actually seen it locally. We definitely don't want to wait 60 seconds before starting the Django server, but we could consider some similar logic to wait for the database in `bin/run`. Feel free to file a new issue for this (though note that iodide is not currently under active development)
